Simple/naive CodeInfo validation pass

jrevels opened the pull request:
This PR implements `validate_code_info(c::CodeInfo)`, a barebones IR validation pass that returns `0` if `c` is valid or a positive integer error code otherwise. I can change the return convention to whatever we want; returning integers was just the easiest way to get started.

The validation pass is in its own file, which is `include`d from `Base.Core.Inference`. Is inference the right location for this? 

I'm also not sure where exactly `validate_code_info` should be called.

ref #22440

jrevels committed:
mock out naive IR validator

jrevels committed:
add more validity constraints

jrevels committed:
implement validate_code_info

jrevels committed:
small bug fixes

jrevels committed:
clarify nslots in docs

vtjnash commented:
> I can change the return convention to whatever we want; returning integers was just the easiest way to get started.

Perhaps `Vector{Exception}`? (The IR is valid if `isempty(errors)`.)

JeffBezanson commented on base/codeinfovalidation.jl:
```diff
@@ -0,0 +1,98 @@
+# This file is a part of Julia. License is MIT: https://julialang.org/license
+
+const VALID_EXPR_HEADS = Symbol[:call, :invoke, :static_parameter, :line, :gotoifnot, :(=),
+                                :method, :const, :null, :new, :return, :the_exception,
```
Another possible rule: `method` is only valid in top-level code (i.e. inside a function with no arguments).

JeffBezanson commented on base/codeinfovalidation.jl:
```diff
@@ -0,0 +1,98 @@
+"""
+function validate_code_info(c::CodeInfo)
+    !(c.inferred) && (c.slottypes != nothing) && return 5
+    (length(c.slotnames) < 1 || c.slotnames[1] != Symbol("#self#")) && return 8
```
The first slot can have a different name for code like `(f::FuncType)(...) = ...`.

JeffBezanson commented on base/codeinfovalidation.jl:
```diff
@@ -0,0 +1,98 @@
+    nslots = length(slotnums)
+    nssavals = length(ssavals)
+    length(c.slotflags) != nslots && return 1
+    length(c.slotnames) != nslots && return 2
```
These checks are too strict. I don't think every slot needs to be referenced. But should check `length(c.slotflags) == length(c.slotnames)`.

JeffBezanson commented on base/codeinfovalidation.jl:
```diff
@@ -0,0 +1,98 @@
+            elseif x.head == :(=) && !(is_valid_lhs(x.args[1]))
+                error_code = 9
+                return true
+            elseif x.head == :call && !(all(is_valid_call_arg(i) for i in x.args[2:end]))
```
`is_valid_call_arg` should also be checked on the argument to `:gotoifnot` exprs.

JeffBezanson commented on base/codeinfovalidation.jl:
```diff
@@ -0,0 +1,98 @@
+
+is_valid_lhs(lhs) = isa(lhs, SlotNumber) || isa(lhs, SSAValue) || isa(lhs, GlobalRef)
+
+is_valid_call_arg(arg) = !(isa(arg, Expr)) || arg.head != :gotoifnot
```
Many more things are invalid here, e.g. LineNumberNode, LabelNode, GotoNode, various other expr heads.

jrevels replied:
Yeah, this was just a start. I'll add the ones you listed.

Maybe someday we can just replace this function with `isa(arg, SSAValue) || isa(arg, SlotNumber)` 😛

JeffBezanson commented:
Cool! This will be nice to have.

I think there should be an additional method for this that accepts a `Method` object, in order to validate that `nargs` and `isva` are consistent with the type signature and CodeInfo.

ararslan commented on base/codeinfovalidation.jl:
```diff
@@ -0,0 +1,98 @@
+    error_code = 0
+    walkast(c.code) do x
+        if isa(x, Expr)
+            if !(in(x.head, VALID_EXPR_HEADS))
```
Super minor point, but parentheses aren't required after `!` when the condition isn't a compound expression of some kind. For consistency with the rest of Base it might be nice to just use `!in(x.head, VALID_EXPR_HEADS)` and likewise elsewhere in here.

jrevels committed:
remove parens from negated conditionals

jrevels committed:
fix some validation conditions

jrevels committed:
return vector of exceptions rather than interger error codes

jrevels commented:
I believe the latest commits implement all of the comments here besides `validate_code_info(::Method)` and the `:method` head check (which I'm going to tackle next). Might be buggy since I haven't written tests yet.

Biggest change so far was @vtjnash's suggestion. Now, `validate_code_info` no longer returns immediately after finding a bug, but keeps chugging along, pushing uncovered bugs into a vector that gets returned at the end. Wasn't sure if performance is a big deal here or not; if so, `CodeInfoError` can be changed to store indices into a preallocated message table instead of taking in a string (assuming those allocations don't already get compiled away - I thought at one point they didn't, but that might be fixed by now).

jrevels committed:
add Method validation on top of CodeInfo validation

jrevels committed:
get started on a couple tests

jrevels committed:
add more code validation tests

jrevels commented:
Alright, added some tests that pass locally (let's see how Travis fares...).

Still have these questions from the OP:

> The validation pass is in its own file, which is `include`d from `Base.Core.Inference`. Is inference the right location for this?

> I'm also not sure where exactly validate_code_info should be called.

jrevels commented:
Anything left to be done here besides hooking up a call to this thing somewhere (presumably somewhere in inference, see my question above)?

Tests pass locally, but I'll rebase to see if we can't get the CI badges green.

tkelman commented on base/codevalidation.jl:
```diff
@@ -0,0 +1,149 @@
+InvalidCodeError(errno, msg) = InvalidCodeError(errno, msg, nothing)
+
+"""
+    validate_code!(errors::Vector{>:InvalidCodeError}, c::CodeInfo)
```
I'm confused, why is this a vector of any supertype of InvalidCodeError?

jrevels replied:
Because the element type can be "anything", as long as I can push objects of type `InvalidCodeError` into it...at least that was my intent, did I screw something up?

tkelman commented on base/codevalidation.jl:
```diff
@@ -0,0 +1,149 @@
+"""
+function validate_code!(errors::Vector{>:InvalidCodeError}, m::Method)
+    if length(m.sig.parameters) != m.nargs
+        push!(errors, InvalidCodeError(11, "number of types in method signature does not match number of arguments",  (length(m.sig.parameters), m.nargs)))
```
wrap long lines

tkelman commented on base/codevalidation.jl:
```diff
@@ -0,0 +1,149 @@
+            push!(errors, InvalidCodeError(7, "not all SSAValues in AST have a type in ssavaluetypes", missing))
+        end
+    else
+        if c.slottypes != nothing
```
`!==`

tkelman commented on test/inference.jl:
```diff
@@ -1102,3 +1102,125 @@ let isa_tfunc = Core.Inference.t_ffunc_val[
+
+# InvalidCodeError 13: encountered Expr head `:method` in non-top-level code
+
+# TODO: This is a tough case to test an isolation...
```
in isolation

jrevels committed:
minor touch-ups

JeffBezanson commented on base/codevalidation.jl:
```diff
@@ -0,0 +1,159 @@
+- `h != :method` for any subexpression head `h` if `m.nargs > 0`
+"""
+function validate_code!(errors::Vector{>:InvalidCodeError}, m::Method)
+    if length(m.sig.parameters) != m.nargs
```
Need `Base.unwrap_unionall(m.sig).parameters`.

JeffBezanson commented on base/codevalidation.jl:
```diff
@@ -0,0 +1,159 @@
+        msg = "number of types in method signature does not match number of arguments"
+        push!(errors, InvalidCodeError(11, msg, (length(m.sig.parameters), m.nargs)))
+    end
+    if m.isva != (last(m.sig.parameters) <: Vararg{Any})
```
This should use `Base.isvatuple`. `Vararg` is not a first-class type and this check will become an error eventually.

vtjnash replied:
These two tests aren't quite correct. If `isva` is true, at most you can say about `length(unwrap_unionall(m.sig.parameters))` is that it is `>= m.nargs - 1`. Otherwise, it is true that they must be exactly equal.

JeffBezanson commented on base/codevalidation.jl:
```diff
@@ -0,0 +1,159 @@
+            if !in(x.head, VALID_EXPR_HEADS)
+                push!(errors, InvalidCodeError(1, "encountered invalid expression head", x.head))
+            elseif x.head == :(=) && !is_valid_lhs(x.args[1])
+                push!(errors, InvalidCodeError(2, "encountered invalid LHS value", x.args[1]))
```
`is_valid_call_arg` can be renamed `is_valid_rvalue`, and also used to check the RHS of assignments here.

JeffBezanson commented on base/codevalidation.jl:
```diff
@@ -0,0 +1,159 @@
+- all assigned-to slots have bit flag 2 set in their respective slotflags
+"""
+function validate_code!(errors::Vector{>:InvalidCodeError}, c::CodeInfo)
+    ssavals = SSAValue[]
```
Maybe this should be a `Set`?

JeffBezanson commented:
I don't feel a strong need to call this anywhere by default (except maybe in a debug build?). More stuff that could potentially slow down the compiler further is the last thing we need.

jrevels committed:
implement suggestions from Jeff and Jameson

jrevels commented:
Okay, I did a refactor to implement the suggestions here (thanks for the review!) and some suggestions Jameson gave me offline. The tests are only half-updated to reflect the new changes; I'll update them/add more tests tomorrow (and probably move them to their own file).

> except maybe in a debug build?

I'm down to add that; what's the right way to do so?

jrevels committed:
enable CodeInfo validation as part of inference for debug builds

jrevels committed:
update CodeInfo validation tests and move them to their own file

jrevels commented:
Took a stab at enabling the validation pass for debug builds. If enabled, it just prints any encountered invalidities as warnings; we can switch to throwing errors later if we want. Do we test debug builds as part of CI?

Also, I updated the tests so that they pass locally and moved them to their own file.

jrevels committed:
do not use warn in Inference since it doesn't exist yet

tkelman commented:
The test looks written in such a way that it would pass even when the code validation isn't enabled by default, right? So should the new file be added to choosetests.jl so it runs? Would it also output warnings if you happened to run the tests with julia-debug? Maybe that's not too bad, there are a few tests that are kind of noisy with warnings already even on release julia (though we should suppress more of those via `@test_warn` over time).

(minor since these aren't exported, but docstrings should generally be imperative)

jrevels committed:
add codevalidation to choosetests.jl

jrevels committed:
use imperative voice for codevalidation docstrings

jrevels commented:
> The test file looks written in such a way that it would pass even when the code validation isn't enabled by default, right? So should the new file be added to choosetests.jl so it runs?

> (minor since these aren't exported, but docstrings should generally be imperative)

Yup. Done and done!

> Would it also output warnings if you happened to run the tests with julia-debug?

Well, it would output warnings whenever the compiler emits "invalid" IR (not just during tests). A warning from the validator indicates there's either a bug in the validator or a bug in the compiler. I'm `make debug`ing now, so I'll see if it actually catches anything during Base test execution. Should be interesting... 

Do we regularly test debug builds anywhere? I skimmed `.travis.yml`/Travis logs and looks like we don't as part of normal CI, but maybe I missed it.

tkelman commented:
> Do we regularly test debug builds anywhere

Not that I can think of. We should maybe add a buildbot that does it, if someone would check it occasionally

tkelman commented:
Are so many warnings expected during debug sysimg?

jrevels committed:
fix validator bugs

jrevels commented:
> Are so many warnings expected during debug sysimg?

The validator had incorrect bounds settings for a couple Expr heads, should be fixed now (at least the ones that I saw).

JeffBezanson commented on base/codevalidation.jl:
```diff
@@ -0,0 +1,166 @@
+# This file is a part of Julia. License is MIT: https://julialang.org/license
+
+# Expr head => argument count bounds
+const VALID_EXPR_HEADS = Pair{Symbol,UnitRange{Int}}[
```
Should this be a `Dict`?

jrevels replied:
It originally was, before I realized that `Dict` isn't defined yet. `Core.Inference` is a strange, strange world 😛

JeffBezanson replied:
Oh, right. `ObjectIdDict` should work though.

jrevels committed:
use ObjectIdDict for VALID_EXPR_HEADS

jrevels committed:
fix overly strict nargs validation check

jrevels committed:
fix typo

jrevels committed:
better message for invalid code warning

jrevels committed:
fix varargs validation bug + tests

jrevels commented:
Finally got around to running the full test suite on the debug build. After fixing some bugs, the only `InvalidCodeError` warnings I got were all `encountered Expr head :method in non-top-level code (i.e. nargs > 0)`.

It seems like a bunch of these warnings were triggered by tests like [this one](https://github.com/JuliaLang/julia/blob/518143fe939c24272fe56d1707604bd50e687cd6/test/core.jl#L3932):

```julia
# issue #15809 --- TODO: this code should be disallowed
function f15809()
    global g15809
    g15809(x::T) where {T} = T
end
f15809()
@test g15809(2) === Int
```

Given the TODO there, I'm assuming the validator was actually correct? Might be good for somebody to give that constraint one last check.

Otherwise, I think this is good to go.

jrevels commented:
Given that tests are passing and it's been a few days, I'll plan on merging this tomorrow (barring anybody finding any other issues).
